### ISSUE TITLE
scikit-learn

### ISSUE THREAD

**Issue by susantkumarsahoo** [susantkumarsahoo/StudentPerformancePrediction-End-To-End#5]
<html>
<body>
<html><head></head><body><h1>Scikit-learn Complete Reference Guide</h1>
<p>A comprehensive reference guide for all major techniques and algorithms available in scikit-learn.</p>
<h2>Table of Contents</h2>
<ul>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#supervised-learning">Supervised Learning</a>
<ul>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#classification-algorithms">Classification Algorithms</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#regression-algorithms">Regression Algorithms</a></li>
</ul>
</li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#unsupervised-learning">Unsupervised Learning</a>
<ul>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#clustering">Clustering</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#dimensionality-reduction">Dimensionality Reduction</a></li>
</ul>
</li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#preprocessing-techniques">Preprocessing Techniques</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#feature-selection">Feature Selection</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#model-selection--validation">Model Selection &amp; Validation</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#ensemble-methods">Ensemble Methods</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#pipeline--composition">Pipeline &amp; Composition</a></li>
<li><a href="https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#metrics">Metrics</a></li>
</ul>
<hr>
<h2>Supervised Learning</h2>
<h3>Classification Algorithms</h3>

Algorithm | Description
-- | --
Logistic Regression | Linear model for binary/multi-class classification
Decision Tree Classifier | Tree-based splits on features
Random Forest Classifier | Ensemble of decision trees
Gradient Boosting Classifier | Sequential tree boosting
AdaBoost Classifier | Adaptive boosting weak learners
XGBoost Classifier | Extreme gradient boosting (if installed)
Support Vector Classifier (SVC) | Finds optimal hyperplane separation
K-Nearest Neighbors (KNN) | Classifies by nearest neighbors voting
Naive Bayes | Probabilistic classifier using Bayes theorem
Neural Network (MLP Classifier) | Multi-layer perceptron
SGD Classifier | Linear classifier with SGD training
Extra Trees Classifier | Extremely randomized trees

<hr>
<h2>Installation</h2>
<pre><code class="language-bash">pip install scikit-learn
</code></pre>
<h2>Usage Example</h2>
<pre><code class="language-python">from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Scale features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train model
model = RandomForestClassifier()
model.fit(X_train_scaled, y_train)

# Predict and evaluate
y_pred = model.predict(X_test_scaled)
accuracy = accuracy_score(y_test, y_pred)
</code></pre>
<hr>
<h2>Resources</h2>
<ul>
<li><a href="https://scikit-learn.org/">Official Documentation</a></li>
<li><a href="https://scikit-learn.org/stable/user_guide.html">User Guide</a></li>
<li><a href="https://scikit-learn.org/stable/modules/classes.html">API Reference</a></li>
</ul>
<hr>
<h2>License</h2>
<p>Scikit-learn is distributed under the BSD 3-Clause license.</p>
<hr>
<p><strong>Last Updated:</strong> October 2025</p></body></html>
</body>
</html># Scikit-learn Complete Reference Guide

A comprehensive reference guide for all major techniques and algorithms available in scikit-learn.

## Table of Contents
- [[Supervised Learning](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#supervised-learning)](#supervised-learning)
  - [[Classification Algorithms](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#classification-algorithms)](#classification-algorithms)
  - [[Regression Algorithms](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#regression-algorithms)](#regression-algorithms)
- [[Unsupervised Learning](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#unsupervised-learning)](#unsupervised-learning)
  - [[Clustering](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#clustering)](#clustering)
  - [[Dimensionality Reduction](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#dimensionality-reduction)](#dimensionality-reduction)
- [[Preprocessing Techniques](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#preprocessing-techniques)](#preprocessing-techniques)
- [[Feature Selection](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#feature-selection)](#feature-selection)
- [[Model Selection & Validation](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#model-selection--validation)](#model-selection--validation)
- [[Ensemble Methods](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#ensemble-methods)](#ensemble-methods)
- [[Pipeline & Composition](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#pipeline--composition)](#pipeline--composition)
- [[Metrics](https://claude.ai/chat/030c831c-8e85-49df-9994-e27331b61f94#metrics)](#metrics)

---

### Classification Algorithms

| Algorithm | Description |
|-----------|-------------|
| **Logistic Regression** | Linear model for binary/multi-class classification |
| **Decision Tree Classifier** | Tree-based splits on features |
| **Random Forest Classifier** | Ensemble of decision trees |
| **Gradient Boosting Classifier** | Sequential tree boosting |
| **AdaBoost Classifier** | Adaptive boosting weak learners |
| **XGBoost Classifier** | Extreme gradient boosting (if installed) |
| **Support Vector Classifier (SVC)** | Finds optimal hyperplane separation |
| **K-Nearest Neighbors (KNN)** | Classifies by nearest neighbors voting |
| **Naive Bayes** | Probabilistic classifier using Bayes theorem |
| **Neural Network (MLP Classifier)** | Multi-layer perceptron |
| **SGD Classifier** | Linear classifier with SGD training |
| **Extra Trees Classifier** | Extremely randomized trees |

### Regression Algorithms

| Algorithm | Description |
|-----------|-------------|
| **Linear Regression** | Fits linear relationship to data |
| **Ridge Regression** | Linear regression with L2 regularization |
| **Lasso Regression** | Linear regression with L1 regularization |
| **Elastic Net** | Combines L1 and L2 regularization |
| **Decision Tree Regressor** | Tree-based prediction |
| **Random Forest Regressor** | Ensemble of regression trees |
| **Gradient Boosting Regressor** | Sequential boosting for regression |
| **Support Vector Regressor (SVR)** | SVM for regression tasks |
| **KNN Regressor** | Predicts by averaging neighbors |
| **MLP Regressor** | Neural network for regression |
| **SGD Regressor** | Linear regression with SGD |

---

### Clustering

| Algorithm | Description |
|-----------|-------------|
| **K-Means** | Partitions data into K clusters |
| **DBSCAN** | Density-based clustering |
| **Hierarchical Clustering** | Creates cluster hierarchy/dendrogram |
| **Mean Shift** | Finds density modes |
| **Spectral Clustering** | Uses graph theory for clustering |
| **Gaussian Mixture Models (GMM)** | Probabilistic clustering |
| **Affinity Propagation** | Message-passing clustering |
| **Birch** | Memory-efficient hierarchical clustering |

### Dimensionality Reduction

| Technique | Description |
|-----------|-------------|
| **PCA** | Principal component analysis, linear projection |
| **t-SNE** | Non-linear visualization technique |
| **Truncated SVD** | Sparse matrix dimensionality reduction |
| **NMF** | Non-negative matrix factorization |
| **LDA** | Linear discriminant analysis |
| **Isomap** | Manifold learning with geodesic distances |
| **Locally Linear Embedding (LLE)** | Local neighborhood preservation |
| **UMAP** | Uniform manifold approximation (if installed) |

---

### Scaling/Normalization

| Technique | Description |
|-----------|-------------|
| **StandardScaler** | Standardizes features (mean=0, std=1) |
| **MinMaxScaler** | Scales features to [0,1] range |
| **RobustScaler** | Scales using median and IQR |
| **Normalizer** | Scales individual samples to unit norm |
| **MaxAbsScaler** | Scales by maximum absolute value |

### Encoding

| Technique | Description |
|-----------|-------------|
| **LabelEncoder** | Encodes labels to integers |
| **OneHotEncoder** | Creates binary columns for categories |
| **OrdinalEncoder** | Encodes ordinal categories to integers |
| **TargetEncoder** | Encodes using target statistics |

### Feature Engineering

| Technique | Description |
|-----------|-------------|
| **PolynomialFeatures** | Generates polynomial combinations |
| **Binarizer** | Converts values to binary |
| **FunctionTransformer** | Applies custom functions |
| **PowerTransformer** | Makes data more Gaussian |

---

## Feature Selection

| Technique | Description |
|-----------|-------------|
| **SelectKBest** | Selects top K features by score |
| **SelectPercentile** | Selects top percentile of features |
| **RFE (Recursive Feature Elimination)** | Recursively removes weakest features |
| **SelectFromModel** | Selects based on model importance |
| **VarianceThreshold** | Removes low-variance features |
| **chi2** | Chi-squared test for categorical features |
| **f_classif** | ANOVA F-test for classification |
| **mutual_info_classif** | Mutual information for classification |

---

## Model Selection & Validation

| Technique | Description |
|-----------|-------------|
| **train_test_split** | Splits data into train/test sets |
| **cross_val_score** | K-fold cross-validation scoring |
| **GridSearchCV** | Exhaustive hyperparameter search |
| **RandomizedSearchCV** | Random hyperparameter sampling |
| **StratifiedKFold** | Stratified K-fold splitting |
| **TimeSeriesSplit** | Time series cross-validation |

---

## Ensemble Methods

| Method | Description |
|--------|-------------|
| **Bagging** | Bootstrap aggregating |
| **Voting Classifier/Regressor** | Combines multiple models by voting |
| **Stacking** | Meta-model on top of base models |

---

## Pipeline & Composition

| Tool | Description |
|------|-------------|
| **Pipeline** | Chains preprocessing and modeling steps |
| **ColumnTransformer** | Applies different transforms to different columns |
| **FeatureUnion** | Combines multiple feature extractors |

---

### Classification Metrics

| Metric | Description |
|--------|-------------|
| **accuracy_score** | Proportion of correct predictions |
| **precision_score** | True positives / predicted positives |
| **recall_score** | True positives / actual positives |
| **f1_score** | Harmonic mean of precision and recall |
| **roc_auc_score** | Area under ROC curve |
| **confusion_matrix** | True/false positive/negative counts |

### Regression Metrics

| Metric | Description |
|--------|-------------|
| **mean_squared_error (MSE)** | Average squared error |
| **mean_absolute_error (MAE)** | Average absolute error |
| **r2_score** | Coefficient of determination |

### Clustering Metrics

| Metric | Description |
|--------|-------------|
| **silhouette_score** | Measures cluster separation |
| **davies_bouldin_score** | Cluster similarity ratio |
| **calinski_harabasz_score** | Variance ratio criterion |

---

## Installation

```bash
pip install scikit-learn
```

## Usage Example

```python
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Scale features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train model
model = RandomForestClassifier()
model.fit(X_train_scaled, y_train)

# Predict and evaluate
y_pred = model.predict(X_test_scaled)
accuracy = accuracy_score(y_test, y_pred)
```

---

## Resources

- [[Official Documentation](https://scikit-learn.org/)](https://scikit-learn.org/)
- [[User Guide](https://scikit-learn.org/stable/user_guide.html)](https://scikit-learn.org/stable/user_guide.html)
- [[API Reference](https://scikit-learn.org/stable/modules/classes.html)](https://scikit-learn.org/stable/modules/classes.html)

---

## License

Scikit-learn is distributed under the BSD 3-Clause license.

---

**Last Updated:** October 2025
